### PR TITLE
Spawn point history

### PR DESCRIPTION
**Pull request by snapper67**
Show stats for all previous spawns at selected spawn point
## Description

Spawns are indicated by a leaf icon. Clicking on the stats link in the spawn info window loads the image and count of all spawns at the location.
## Motivation and Context

http://feathub.com/PokemonGoMap/PokemonGo-Map/+11
## How Has This Been Tested?

Local Testing
## Screenshots (if appropriate):
screen shot 2016-08-14 at 3 13 12 pm

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: pogom/models.py [hunk 1]
```diff
@@ -204,6 +204,21 @@ def get_spawnpoints(cls, swLat, swLng, neLat, neLng):
...
+    @classmethod
+    def get_spawn_data(cls, spawnpoint_id):
+        query = (Pokemon
+                 .select(fn.Count(Pokemon.pokemon_id).alias('count'), Pokemon.pokemon_id)
+                 .where((Pokemon.spawnpoint_id == spawnpoint_id))
+                 .group_by(Pokemon.pokemon_id)
+                 .order_by(Pokemon.pokemon_id)
+                 .dicts())
+
+        spwawn_data = []
```
Comment (chuyskywalker) [inline pogom/models.py]: `spwawn` data? :)

**Comment by chuyskywalker** [comment]
Why not just load the spawn data upon opening the info pane?

**Comment by Xcelled** [comment]
Nitpicking, but 'history' instead of 'data' would be better. Also, this is going to suffer from the same issue as #611

File: pogom/models.py [hunk 2]
```diff
@@ -204,6 +204,21 @@ def get_spawnpoints(cls, swLat, swLng, neLat, neLng):
...
+        for p in query:
```
Comment (Xcelled) [inline pogom/models.py]: Get rid of this. Just do `return list(query)`.

**Comment by snapper67** [comment]
Loading spawn data on opening the window pane would send a request on every spawn point that was moused over.

**Comment by snapper67** [comment]
@Xcelled  I ran in sqlite with no problems and manually ran the query on a MySql 5.7.13 database with sql_mode set to `ONLY_FULL_GROUP_BY` with no errors. I use a MySql 5.6.24 on the computer that runs the map.

**Comment by G1K** [comment]
It would be very useful figures in minutes when it appears spawnpoint pokemon sectional hours.

**Comment by G1K** [comment]
Perhaps sorting is done on the number? Uncomfortable as it is to look for someone spawn. And also nice to do by means of javascript count percentage REP. such as suggested here

http://feathub.com/PokemonGoMap/PokemonGo-Map/+35

**Comment by j16sdiz** [comment]
1. Clicking the "stat" link twice screw up the UI.

2. Adding spawn time for that spawnpoint would be nice.

**Comment by snapper67** [comment]
@G1K The circles were hard to hover over or click on to get the info window

**Comment by fasteX89** [comment]
Wow nice work spawntime would be really nice

**Comment by Xcelled** [comment]
To show spawn times accurately, the Pokemon table's PK needs to be altered. I did this in #589. If no one has a better idea than using the 'despawn' time, #589 could be updated and reopened to allow accurate collecting of time.

**Comment by fasteX89** [comment]
would be awesome for prediction of the next spawn at that spawnpoint

**Comment by Zentriert** [comment]
I love the idea of this, there's a good bit of lag trying to bring up the heat map for spawns from the stats page so this would be helpful and there's a few very small "nests" of some sort by me that always spawn rare pokemon so it'd be great to check just their spawn points history.

Also the stats page only shows the last 24 hours .. anyway to change that, see more?

**Comment by EmpowerZ** [comment]
Can we have an option to turn those leaves into most common pokemon for the spawn? This way we could find nests / common places for a pokemon much easier.

**Comment by xrobau** [comment]
@EmpowerZ that seems like a good idea, but how would you render it so it's obvious it's not a 'live' spawn?

**Comment by EmpowerZ** [comment]
@xrobau I thought this gonna be a separate page for these stats. And not overloading the main page with it. 
This way there is no need to to render it differently from live spawns

**Comment by vinsilsim** [comment]
Marge please!!!!!!!!!!!!

**Comment by acidflare1** [comment]
I have a suggestion on how you can display the information. How about a 'Spawn' link on the Stats menu? Then you can see each individual spawn point to select it. It will list the GPS coordinates and also show a piece of the map where the coordinates are located(so as not to be confused with what may seem like an arbitrary number). If there are multiple points on the map it can be highlighted with a different color to differentiate it. It will look similar to this:
spawn point selection
Then after you've clicked the 'Spawn History' you'll see something similar to this:
spawn point history
The map search feature can run the way it has been. It can just take already existing data from the db file and organize it so you can see if there's patterns for certain points. It'll also allow you to see if that particular spawn point is worth checking in the future.

**Comment by snapper67** [comment]
I have considered loading it a panel like the one that the stats are loaded into. It would slide out when you click on the stats button. Then there could be a lot more stats in that area.

@EmpowerZ This part of the main screen and was integrated with a the spawn code that was checked in after I started this. I did consider replacing the leaf with the minute that the spawn most often occurs.

**Comment by tim95030** [comment]
This feature seems to be working from my testing for the last couple days.

**Comment by NoobDev3** [comment]
I haven't pulled  this in a bit, so it might've changed already. Can you put the stats when you first click the spawn point? Instead of having the stats link after opening the spawn.

**Comment by beccasafan** [comment]
I think this is a lot of content to show in an info window and that it would be better served in the stats page and with stat overlays. If #969 gets merged in these could work together and that info window could link to the stat overlays with the pokemon spawn history for that spawn point. Best of both worlds?

**Comment by vinsilsim** [comment]
Marge please!

**Comment by Bart274** [comment]
how far is this feature from being merged?

**Comment by slideup-benni** [comment]
It seems like snapper67 is away :disappointed:

**Comment by snapper67** [comment]
If I thought this was going to be merged, I would keep it up to date. I have this merged with my local copy, but keeping the branch up to date started to feel like a waste of time. I also have a history window that shows everything that has spawned in the current area of the map.
 The screenshot shows that in this are Voltorbs are the most common spawn.

screen shot 2016-09-03 at 10 38 59 am

**Comment by markvl91** [comment]
I'd love this!

**Comment by Bart274** [comment]
keep this updated, I hope this gets merged :)

**Comment by aldarund** [comment]
+1

**Comment by jobasbdo** [comment]
Love this, merge pls

**Comment by Saprophator** [comment]
Please update @snapper67, would love to see this merged.
maybe invisiblek or @chuyskywalker can say what is keeping this from getting merged!?

**Comment by nolageek** [comment]
Seriously would love to see this merged as well. There needs to be some way to better iolate what spawns where in the stats page - especially when scanning multiple locations. This goes a long way to do that.

**Comment by justsomedudeonthenet** [comment]
This is a feature I would love to see added.  Please fix the merge conflicts so that it works with the new spawnpoint indicators.

**Comment by bohigunm** [comment]
This is a great feature

**Comment by slideup-benni** [comment]
@snapper67 i'm not a mod, but i think that you have to clear the branch conflicts before a mod will merge this. With this conflicts nobody can test your work with the current state, so no mod can say yes or no. Its a vicious circle...

**Comment by nolageek** [comment]
Excellent - hopefully we can get this merged. :)

File: pogom/models.py [hunk 3]
```diff
@@ -327,6 +327,17 @@ def get_spawnpoints_in_hex(cls, center, steps):
 
         return filtered
 
+    @classmethod
+    def get_spawn_history(cls, spawnpoint_id):
```
Comment (aldarund) [inline pogom/models.py]: What the meaning of use classmethod here?

**Comment by nolageek** [comment]
What would be the prefered way to add this to my local clone? 'git cherry-pick'?

**Comment by nowrasahole** [comment]
hi i merged this development branch commit1004, i have had a problem with the images for pokemon not showing up at gyms with -gi enabled since it was introduced. 

Now i have the same problem with the spawn point stats page. i can click on stat and see the numbers but no icons. 

This wasn't a major problem with the gyms because when you hover over the icon it tells you the name of the pokemon.   Windows 10 x64 with maria db mysql.

**Comment by djgreg13** [comment]
```
url: '/spawn_history?spawnpoint_id=' + spawnpointId
```

please fix this, some people can run instance in another than /

**Comment by drallieiv** [comment]
Will this override the current coloring for next spawns ?

**Comment by primaxius** [comment]
@snapper67 Hey just reviewing your PR ...just wanted to make sure if this is supposed to include https://github.com/PokemonGoMap/PokemonGo-Map/pull/669#issuecomment-244550153 or not as I couldnt see this in the PR currently.

**Comment by snapper67** [comment]
This is only the spawn history when selecting a spawn point. The comment shows https://github.com/PokemonGoMap/PokemonGo-Map/pull/571 which gives the history for all spawns in the view area. That one would need to be redone with all the changes since it was created. So only spawn history in this one

**Comment by primaxius** [comment]
Declined - this much information is best suited for a stats page of sorts.

**Comment by Bart274** [comment]
Why is this declined? I was really looking forward to this! I am really disappointed that this isn't coming! :(

Op 26 sep. 2016 om 20:07 heeft primaxius <notifications@github.com<mailto:notifications@github.com>> het volgende geschreven:


Declined - this much information is best suited for a stats page of sorts.

-

**Comment by primaxius** [comment]
Ok let me be clear here. The idea is great and we do need this however this implementation using the info window will be a no-go as this really clutters a very small window. If the coder has changes he can msg me an i'll reopen it. or even resubmit as it would be a big change from this PR.

**Comment by Bart274** [comment]
I see :)
Yeah that's better :)

What would also be nice if you could filter the full stats page between all locations or the maps current location

Op 26 sep. 2016 om 20:23 heeft primaxius <notifications@github.com<mailto:notifications@github.com>> het volgende geschreven:


Ok let me be clear here. The idea is great and we do need this however this implementation using the info window will be a no-go as this really clutters a very small window. If the coder has changes he can msg me an i'll reopen it. or even resubmit as it would be a big change from this PR.

-

**Comment by mew905** [comment]
I'd really like to see what AcidFlare1 posted as well.

**Comment by kenjif70** [comment]
Just FYI, this is how one of popular Japanese maps is doing.

image

**Comment by snapper67** [comment]
If I get a chance I will redo this loading it into a side panel like stats does

**Comment by acidflare1** [comment]
What if instead it's a completely separate window? It'll open a separate browser window that will make references to the db file just to display the data. Could be a separate program altogether to view the data but not tack on more processes to the scanner. That way you can view historical data with out having to run the scanner. Just in case the scanner becomes no longer viable you'll still be able to use the history.

**Pull request closed without merging**